Create basic database

zfrank7777 opened the pull request:
- Create SQLite database
- Build User class in models
- Offer basic register api (api_register)
- Not connect to the database yet (no session related operation)

zfrank7777 committed:
Create basic database

- Create SQLite database
- Build User class in models
- Offer basic register api (api_register)
- Not connect to the database yet (no session related operation)

amjltc295 commented on src/surfing_penguin/routes.py:
```diff
@@ -50,6 +51,18 @@ def post(self):
+        password = request.form['password']
+        user = User(username, password)
+        return "user: %s, passwd:%s" % (user.username, user.password)
+
```
You need to provide a test for this function, either a simple html or unit test.

amjltc295 commented on src/surfing_penguin/models.py:
```diff
@@ -1,13 +1,26 @@
+    id = Column(Integer, primary_key=True)
+    username = Column(String)
+    password = Column(String)
+
```
Should add some columns to record register time, last login time, etc.

amjltc295 reviewed (commented):
Overall looks good, but I have no way to test your new post function.

zfrank7777 committed:
Build some user account associated APIs.

Add login_time and register_time to User model.
Add APIs: register, login, delete_user, search_user.
(The programm is now interacting with SQLite database)

zfrank7777 committed:
Update user's login time when login.

Add register time and last login time in search_user.
Update user's login time when login.

zfrank7777 committed:
Complete basic operations on User.

Add convert_user_to_json functions.
Finish show_users function.

amjltc295 commented on src/surfing_penguin/routes.py:
```diff
@@ -40,27 +41,76 @@ def get(self):
+        if session.query(User).filter_by(username=new_name).count() == 1:
+            return "Please use another username"
+        new_user = User(new_name, new_password)
+        session.add(new_user)
```
These operations need to be moved to db_interface

amjltc295 commented on src/surfing_penguin/routes.py:
```diff
@@ -40,27 +41,76 @@ def get(self):
+class register(Resource):
+    def get(self):
+        return "register: enter username and password"
+
```
Need API models for these APIs

amjltc295 commented on src/surfing_penguin/routes.py:
```diff
@@ -40,27 +41,76 @@ def get(self):
+            return "user does not exist"
+        user = session.query(User).filter_by(username=search_name).first()
+        return "user %s, register time: %s, last login: %s" % (
+                user.username, user.register_time, user.last_login)
```
Use "{}, {}".format(a, b) instead of %s

amjltc295 reviewed (changes requested):
Needs some modification

zfrank7777 committed:
Do the revision suggested above.

- Move all session-related function to db_interface.
- Modify the return format.

zfrank7777 committed:
Update requirements.txt

amjltc295 commented on src/surfing_penguin/db_interface.py:
```diff
@@ -1,4 +1,38 @@
+from surfing_penguin import api, session
+from surfing_penguin.models import User
+
+
```
Ther User_func should be named as UserFunctions. Please take a look at https://www.python.org/dev/peps/pep-0008/

amjltc295 commented on src/surfing_penguin/db_interface.py:
```diff
@@ -1,4 +1,38 @@
+    def delete_user(name):
+        session.query(User).filter_by(username=name).delete()
+        session.commit()
 
```
The above parts look good.

amjltc295 commented on src/surfing_penguin/routes.py:
```diff
@@ -1,19 +1,26 @@
         'id': fields.Integer
     })
 
 
```
Descriptions for API models should be added. Also, API models for POST functions are missing.

amjltc295 commented on src/surfing_penguin/routes.py:
```diff
@@ -47,69 +54,53 @@ def get(self):
+        if new_user:
+            return "registratoin done for {}".format(new_user.username)
+        return "register failed"
 
```
Use API model instead of strings.

amjltc295 reviewed (changes requested):
The db_interface looks good. API layer needs some modification. Also, testing methods (unit tests or at least curl scripts) should be included.

zfrank7777 committed:
Build api models for users.

amjltc295 reviewed (commented):
Good for having all API testable in the browser.
However, there should be more API models for different
user operation and password should not be send back in any way.

zfrank7777 committed:
Distinguish input and output user models.

amjltc295 reviewed (approved):
lgtm
